src/Datadog.Trace/Logging/CustomNLogLogProvider.cs

```diff
@@ -43,23 +43,48 @@ protected override OpenMdc GetOpenMdcMethod()
 
             var mdcContextType = FindType("NLog.MappedDiagnosticsContext", "NLog");
             var setMethod = mdcContextType.GetMethod("Set", typeof(string), typeof(object));
-            var removeMethod = mdcContextType.GetMethod("Remove", typeof(string));
-            var valueParam = Expression.Parameter(typeof(object), "value");
-            var setMethodCall = Expression.Call(null, setMethod, keyParam, valueParam);
-            var removeMethodCall = Expression.Call(null, removeMethod, keyParam);
 
-            var set = Expression
-                .Lambda<Action<string, object>>(setMethodCall, keyParam, valueParam)
-                .Compile();
-            var remove = Expression
-                .Lambda<Action<string>>(removeMethodCall, keyParam)
-                .Compile();
+            if (setMethod != null)
+            {
+                var removeMethod = mdcContextType.GetMethod("Remove", typeof(string));
+                var valueParam = Expression.Parameter(typeof(object), "value");
+                var setMethodCall = Expression.Call(null, setMethod, keyParam, valueParam);
+                var removeMethodCall = Expression.Call(null, removeMethod, keyParam);
+
+                var set = Expression
+                    .Lambda<Action<string, object>>(setMethodCall, keyParam, valueParam)
+                    .Compile();
+                var remove = Expression
+                    .Lambda<Action<string>>(removeMethodCall, keyParam)
+                    .Compile();
 
-            return (key, value, _) =>
+                return (key, value, _) =>
+                {
+                    set(key, value);
+                    return new DisposableAction(() => remove(key));
+                };
+            }
+            else
             {
-                set(key, value);
-                return new DisposableAction(() => remove(key));
-            };
+                setMethod = mdcContextType.GetMethod("Set", typeof(string), typeof(string));
+                var removeMethod = mdcContextType.GetMethod("Remove", typeof(string));
+                var valueParam = Expression.Parameter(typeof(string), "value");
+                var setMethodCall = Expression.Call(null, setMethod, keyParam, valueParam);
+                var removeMethodCall = Expression.Call(null, removeMethod, keyParam);
+
+                var set = Expression
+                    .Lambda<Action<string, string>>(setMethodCall, keyParam, valueParam)
+                    .Compile();
+                var remove = Expression
+                    .Lambda<Action<string>>(removeMethodCall, keyParam)
+                    .Compile();
+
+                return (key, value, _) =>
+                {
+                    set(key, value.ToString());
```

Comment:
I wonder if we should just use this one all the time? 🤔

Reply:
That sounds fine to me. The only values we pass to this method are string objects anyways, so the impact of the added `ToString()` should be minimal

Reply:
🤦🏼 we can't invoke `ToString()` here because we would immediately call `ToString()` on our `LogEnricher` object when we try to add it at the beginning of the trace. NLog needs the exact string representation when the API is called so I think the fallback is to actually avoid using our optimized `LogEnricher` in this case. I'll need to re-do the PR